### ISSUE TITLE
Lifecycle methods called on objects of Components that use member_name assembly can result in unexpected behavior.

### ISSUE THREAD

**Issue by mzipay** [mzipay/Aglyph#5]
One of the object creation patterns that Aglyph supports is "member_name" assembly, where an object of a component is not so much "assembled" as it is "acquired" via attribute access on a module or class.

But in these cases, the object is truly a singleton (in the Python sense, not just the Aglyph sense). This means that even if multiple components with _different_ IDs use the same member, it is possible that the after_inject and/or before_clear lifecycle methods may be called more than once on the same object. If those lifecycle methods modify the state of the object, then this could lead to unexpected behavior in an application.

Here is an example:

```
# module.py
class MyClass:
    def __init__(self):
        self.state = 1
    def transition(self):
        self.state *= 2

MY_OBJ = MyClass()
```

And the corresponding Aglyph context:

```
<component
   id="my-obj-1"
   dotted-name="module"
   member-name="MY_OBJ"
   strategy="singleton"
   after_inject="transition"
/>
<component
   id="my-obj-2"
   dotted-name="module"
   member-name="MY_OBJ"
   after_inject="transition"
/>
```

Now let's say an assembler assembles both "my-obj-1" and "my-obj-2":

```
>>> obj1 = assembler.assemble("my-obj-1")
>>> obj2 = assembler.assemble("my-obj-2")
```

At first glance (or if this were not such a trivial example), we might expect that both `obj1` and `obj2` would be in state 2. But in fact **both obj1 and obj2 will be in state 4**:

```
>>> obj1.state
4
>>> obj2.state
4
```

This is irrespective of the assembly strategy (note that "my-obj-1" is a singleton, while "my-obj-2" is a prototype).

**Comment by mzipay** [comment]
It's tempting to leave this as-is, since pedantically it is "expected" behavior... but practically speaking, I'd like to at least issue a RuntimeWarning when lifecycle methods are specified (explicitly or implicitly) for a component that uses member_name.

Another approach, but more restrictive, would be to enforce mutual exclusivity on member_name and after_inject/before_clear. I'm less in favor of this approach at the moment because it's difficult to state definitively that there is no case, **_ever_**, where this behavior (possibly calling a lifecycle method twice on the same instance) would be desired.

(Aglyph already issues a RuntimeWarning if a lifecycle method raises an exception, but I feel it would be more explicit to communicate the possibility of unexpected runtime behavior if it can be detected.)

**Comment by pboechat** [comment]
I don't know if you should call this component a singleton or a prototype (or anything related to any other strategy). To me it's simply a global name that you're importing into Aglyph's context. You didn't assemble it, and so you shouldn't let the user choose an assembly strategy. I guess "member_name" should implicitly make use of a new strategy called "imported", which would exempt you from taking care of the object's life cycle and should not let the user control caching (with _clear_[...] methods). As for "after_inject" and "before_inject", those are template methods for the injection mechanism and should still be available.

**Comment by mzipay** [comment]
I like the idea of "imported" as a strategy for member_name.  That makes the intent clear, and also makes it much easier to restrict or issue warnings based on usage. Thanks!

**Comment by pshirali** [comment]
While the "imported" dependency would become a component in the context, I'd also like to highlight the possibility of having to _resolve_ the value of the component each time it is injected as a dependency elsewhere. This way, if the value of the "member_name" changes through the lifecycle of the process, objects built during the lifecycle of the process always consume the latest value from that "member_name" (as opposed to the value it had at the time of context resolution/building the assembler, which probably might be at the start of the application's lifecycle)

**Comment by mzipay** [comment]
Agreed, pshirall. It would be an unexpected surprise if the imported dependency were somehow "frozen" to the value it had at initial resolution time. Python itself doesn't work that way, so Aglyph won't either. Thanks for the input.

**Comment by mzipay** [comment]
The "_imported" strategy is implicit (automatically used and only valid for components that specify a member_name). Under the hood, creating "_imported" components is identical to creating "prototype" components, which is directly relevant to pshirall's comment above.
These changes will be included in release 3.0.0.

**Comment by mzipay** [comment]
Running test suites for 3.0.0 currently, and this new behavior is testing well so far, so closing this issue. I anticipate 3.0.0 will be released in the next week or so.
